hr_timesheet_validator/model/hr_timesheet_sheet.py

```diff
@@ -0,0 +1,154 @@
+            elif (
+                timesheet.department_id
+                and timesheet.department_id.parent_id
+                and timesheet.department_id.parent_id.manager_id
+                and timesheet.department_id.parent_id.manager_id.user_id
+                and timesheet.department_id.parent_id.
+                    manager_id.user_id.id != uid
+            ):
+                users[timesheet.id].append(
+                    timesheet.department_id.manager_id.user_id.id)
+
+            [res[timesheet.id].append(item) for item in users[timesheet.id]
+             if item not in res[timesheet.id]]
+        return res
+
+    def button_confirm(self, cr, uid, ids, context=None):
+        validators = self._get_validator_user_ids(cr, uid,
+                                                  ids, context=context)
+        for sheet in self.browse(cr, uid, ids, context=context):
+            self.write(cr, uid, sheet.id,
+                       {'validator_user_ids':
+                           [(4, user_id) for user_id
+                            in validators[sheet.id]]})
+        return super(hr_timesheet_sheet, self).button_confirm(cr, uid, ids,
+                                                              context=context)
+
+    def _check_authorised_validator(self, cr, uid, ids, *args):
+        model_data_obj = self.pool.get('ir.model.data')
+        res_groups_obj = self.pool.get("res.groups")
+        group_hr_manager_id = model_data_obj._get_id(
+            cr, uid, 'base', 'group_hr_manager')
+        group_hr_user_id = model_data_obj._get_id(
+            cr, uid, 'base', 'group_hr_user')
+
+        for timesheet in self.browse(cr, uid, ids):
+            if group_hr_manager_id:
+                    res_id = model_data_obj.read(cr, uid,
+                                                 [group_hr_manager_id],
+                                                 ['res_id'])[0]['res_id']
+                    group_hr_manager = res_groups_obj.browse(
+                        cr, uid, res_id)
+                    group_hr_manager_ids = [user.id for user
+                                            in group_hr_manager.users]
+                    if uid in group_hr_manager_ids:
+                        continue
+
+            if group_hr_user_id:
+                    res_id = model_data_obj.read(cr, uid, [group_hr_user_id],
+                                                 ['res_id'])[0]['res_id']
+                    group_hr_user = res_groups_obj.browse(
+                        cr, uid, res_id)
+                    group_hr_user_ids = [user.id for user
+                                         in group_hr_user.users]
+                    if uid in group_hr_user_ids:
+                        continue
+
+            validator_user_ids = []
+            for validator_user_id in timesheet.validator_user_ids:
+                validator_user_ids.append(validator_user_id.id)
+            if uid not in validator_user_ids:
```

Comment:
Definitely a case for list comprehension: `validator_user_ids = [x.id from x in timesheet.validator_user_ids]`